#20 feat 로그인 - DB 매칭 및 ExceptionHandler 수정

yunah7 opened the pull request:
## 요약
로그인 기능 추가 및 예외 처리 수정


## 변경 내용
- **이전 pr 내용 반영 
   #17** 
  - Controller
     - API 수정
     -  valid -> validated 어노테이션

  - Service
    - BCrypt 함수를 static으로 import 

- **로그인 기능**
  - #14 에서 추가한 getUser 메소드를 통해 회원 존재 여부 검사
  - Boolean 타입의 activate를 UserDTO에도 추가, 계정 활성화 여부 검사
  - int 타입의 loginCount를 UserDTO에도 추가, 로그인 실패 횟수 기록
  - 로그인 성공 시, userInput을 userSaved로 대체
     비밀번호 마지막 변경일로부터 90일이 지났는지 검사 및 로그인 실패 횟수 초기화
  -  로그인 실패 시, 실패 횟수가 4회였으면, 비활성화&실패 횟수 초기화
     그 이하라면, 실패 횟수만 증가 후 예외 throw

- **예외 처리**
  -  예외 종류를 enum 클래스로 묶음 
  -  enum클래스를 생성자로 받고, RuntimeException을 상속받는 클래스 생성


## Issue number/URL
   #20

yunah7 committed:
#20 feat : 로그인 - DB 매칭/ExceptionHandler 수정

 Mapper을 통해 입력값을 매칭해 로그인
 enum 클래스와 RuntimeException 클래스를 상속 받는 클래스를 생성해
 예외처리

sunwootest commented:
out/production이 자꾸 추가되는데 조치 부탁드립니다 :)

sunwootest commented on src/main/java/study/sharedcalendar/controller/UserController.java:
```diff
@@ -1,28 +1,31 @@
+    public void signUp(@RequestBody @Validated UserDTO user) {
+        //if (errors.hasErrors()) {
+        //    throw new ValidationException("INVALID");
+        //}
```
주석은 최소화해주세요 :) 만약 추후에 유지보수가 필요한 코드에 대한 주석이라면 이에 대한 설명을 자세하게 작성해주는 것이 좋아보이네요 :)

sunwootest commented on src/main/java/study/sharedcalendar/dto/UserDTO.java:
```diff
@@ -23,5 +23,8 @@ public class UserDTO {
     private String inviteUrl;
+
+    private boolean activate;
+    private int loginCount;
```
개행 처리가 통일이 안된 느낌입니다. 컨벤션을 정해서 지켜주세요 :)

sunwootest commented on src/main/java/study/sharedcalendar/exception/ExceptionError.java:
```diff
@@ -0,0 +1,15 @@
+        this.httpStatus = errorCode.getHttpStatus();
+        this.message = errorCode.getMessage();
+    }
+}
```
ExceptionError라는 이름이 모든 예외를 처리해줄 것처럼 느껴집니다. 좀 더 구체적인 느낌이 들게 이름을 변경할 필요가 있어보입니다 :)

yunah7 replied:
> 이전 PR에서 리뷰해드린 내용은 이전 PR에서 반영해주세요! 나름 복잡한 요구사항을 잘 처리하셨는데 로그인은 단순 DB 매칭만으로 이루어지지 않습니다. 우리는 세션을 이용해서 로그인을 구현하기로 했는데 이 부분까지 구현해주세요 :)

리뷰 내용을 확인하기 전에 코드를 더 추가해버려 같이 작성하게 되었습니다. 😅 바로 이전에 로컬에서 깃을 사용하려다 작성 중이던 코드를 반쯤 날려버리는 실수를 해서, 해당 이슈는 일단 이렇게 올리게 되었습니다.. 

세션과 관련된 내용은 추가되는 코드가 꽤 있어 이슈를 분리해 작성할 예정입니다!

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -17,24 +19,66 @@ private UserDTO getUser(UserDTO userInput){
+        int userLoginCount = userSaved.getLoginCount();
+        if(checkPassword(userInput.getPassword(), userSaved.getPassword())){
+            userInput = userSaved;
+            if((userMapper.passwordChangedDate(userInput.getId()))>=90){
+                throw new ExceptionError(ErrorCode.PASSWORD_EXPIRED);
+            }
```
90과 같은 내용은 properties파일로 빼놓아도 될 것 같네요. 법이 바뀌어서 30일마다 변경하라고 요구사항이 추가되면 이 클래스를 다 뜯어봐야할 수 있어서요 :)

uri010 replied:
ErrorCode도 import static 해주면 더 깔끔할거 같아요!

yunah7 replied:
수정하겠습니다 !

sunwootest reviewed (commented):
이전 PR에서 리뷰해드린 내용은 이전 PR에서 반영해주세요! 나름 복잡한 요구사항을 잘 처리하셨는데 로그인은 단순 DB 매칭만으로 이루어지지 않습니다. 우리는 세션을 이용해서 로그인을  구현하기로 했는데 이 부분까지 구현해주세요 :)

uri010 commented on src/main/resources/ddl.sql:
```diff
@@ -41,7 +41,7 @@ create table connector (
     activate tinyint not null,
     created_at timestamp not null default current_timestamp,
-    updated_at timestamp not null default current_timestamp on update current_timestamp,
+    updated_at timestamp not null default current_timestamp,
```
default에 on update current_timestamp를 지운 이유가 있을까요? on updated current_timestamp로 해놓으면 수정할 때마다 자동으로 시간이 업데이트 되는 걸로 압니다.

yunah7 committed:
#20 fix : 로그인 - 커스텀property 추가 / 양식 수정

자바 코드 컨벤션을 따른 양식 수정
exception 패키지 내 클래스들의 이름 변경
DB table 규칙 지향 및 에러 해결을 위해 테이블 명 수정
상수로 작성했던 비밀번호 유효 기간을 application.properties에 커스텀 변수로 추가

yunah7 commented:
gitignore 파일에 out 폴더를 추가하였고, 코드 양식 또한 수정했습니다.
DB 테이블명 중, 에러를 피하기 위해 임시로 설정해두었던 users 및 다른 테이블의 이름을 네이밍 규칙에 맞추어 변경했습니다.
상수로 지정해두었던 비밀번호 유효 기간은 properties에 변수를 추가해 사용하는 것으로 바꾸었으며, 
exception 패키지 내 클래스들의 이름도 수정하였습니다.

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -1,40 +1,90 @@
 @RequiredArgsConstructor
 public class UserService {
+
+    @Value("${PASSWORD_EXPIRED_DATE}")
+    private int PASSWORD_EXPIRED_DATE;
```
상수는 final로 선언하는게 나아보이네요. 아마 빈 주입시 문제가 생겨서 이렇게 선언하신 것 같은데 롬복을 사용하지 않는 생성자 주입을 사용하시면 해결됩니다 :)

sunwootest commented on src/main/java/study/sharedcalendar/service/UserService.java:
```diff
@@ -1,40 +1,90 @@
+
+            modifyLoginCount(userSaved.getId(),0);
+        } else{
+            if(userLoginCount==4){
```
이 부분도 properties에서 참조할 수 있을 것 같네요

sunwootest commented on src/main/resources/application.properties:
```diff
@@ -1 +1,15 @@
+spring.h2.console.enabled=true
+spring.h2.console.path=/h2-console
+spring.datasource.url=jdbc:h2:mem:test;DB_CLOSE_DELAY=-1;MODE=MySQL;NON_KEYWORDS=USER
+spring.datasource.password=
+spring.datasource.username=sa
+spring.datasource.driver-class-name=org.h2.Driver
 
+mybatis.mapper-locations=classpath:mybatis-mapper/*.xml
+mybatis.configuration.map-underscore-to-camel-case=true
+spring.sql.init.mode=embedded
+spring.sql.init.schema-locations=classpath:ddl.sql
+
+logging.level.org.apache.coyote.http11=debug
+
+PASSWORD_EXPIRED_DATE=90
```
ignore 적용이 안된 것 같네요. 이 부분이 해결이 잘 안되는 부분이긴 한데 캐시삭제와 같은 다양한 방법이 있으니 해결해보세요!
